src/Countries/Singapore.php

```diff
@@ -0,0 +1,171 @@
+        );
+    }
+
+    /** @return array<string, string|CarbonImmutable> */
+    protected function variableHolidays(int $year): array
+    {
+
+        $easter = $this->easter($year);
+
+        return array_merge(
+            $this->replacementForSunday("Chinese New Year Day 1", $year, $this->chineseCalendar("01-01", $year), 2),
+            $this->replacementForSunday("Chinese New Year Day 2", $year, $this->chineseCalendar("01-02", $year), 1),
+            $this->replacementForSunday("Hari Raya Pausa", $year, $this->islamicCalendar("01-10", $year), 1),
+            $this->replacementForSunday("Vesak Day", $year, $this->chineseCalendar("04-15", $year), 1),
+            $this->replacementForSunday("Hari Raya Haji", $year, $this->islamicCalendar("10-12", $year), 1),
+            $this->getDeepavali($year),
+            [
+                'Good Friday' => $easter->subDays(2),
+            ]
+        );
+    }
+
+    /** @return array<string, string|CarbonImmutable> */
+    protected function replacementForSunday(string $name, int $year, string $input, int $addDays = 0): array
+    {
+        /** @var CarbonImmutable $dateTime */
+        $dateTime = CarbonImmutable::createFromFormat("Y-m-d", $year . "-" . $input);
+        
+        $replacementDay = [];
+        $holiday = [$name => $dateTime->format("m-d")];
+        if ($dateTime->isSunday()) {
+            $dateTime = $dateTime->addDays($addDays);
+            $replacementDay = ["$name (Replacement)" => $dateTime->format("m-d")];
+        }
+
+        return array_merge(
+            $holiday,
+            $replacementDay
+        );
+    }
+
+    /** @return array<string, string|CarbonImmutable> */
+    protected function getDeepavali(int $year): array
+    {
+        // Kartik month
+        $date = "08-01";
+
+        $deepavaliDateBasedFromIndianCalendar = $this->indianCalendar($date, $year);
+
+        $regularHoliday = $this->getPhaseNextNewMoon($year . "-" . $deepavaliDateBasedFromIndianCalendar);
+        return $this->replacementForSunday("Deepavali", $year, $regularHoliday, 1);
+    }
+
+
+    protected function getIslamicFormatter(): IntlDateFormatter
+    {
+        return new IntlDateFormatter(
+            locale: 'en-SG@calendar=islamic-civil',
+            dateType: IntlDateFormatter::MEDIUM,
+            timeType: IntlDateFormatter::NONE,
+            timezone: $this->timezone,
+            calendar: IntlDateFormatter::TRADITIONAL
+        );
+    }
+
+    protected function getHijriYear(int $year, int $nextYear = 0): int
+    {
+        $formatter = $this->getIslamicFormatter();
+        $formatter->setPattern('yyyy');
+
+        /** @var CarbonImmutable $dateTime */
+        $dateTime = CarbonImmutable::createFromFormat("d/m/Y", '01/01/' . ($year + $nextYear));
+
+        $hijriYear = $formatter->format($dateTime);
+
+        return (int) $hijriYear;
+    }
+
+    protected function chineseCalendar(string $input, int $year): string
```

Comment:
I think you could use the ChineseCalendar trait

Reply:
I will review on this.